pkgs/build-support/rust/build-rust-package/default.nix

```diff
@@ -50,7 +49,7 @@ lib.extendMkDerivation {
       cargoDepsHook ? "",
       buildType ? "release",
       meta ? { },
-      useFetchCargoVendor ? false,
+      useFetchCargoVendor ? true,
```

Comment:
I was thinking we'd go a release cycle requiring this to be set, so that it's not possible to silently end up with broken FODs.  Will accidentally keeping a `fetchCargoTarball` hash around result in a broken build, at least?

Reply:
No invalid hashes will go undetected, because a different name will cause FOD hashes to be invalidated.

Reply:
Right, the risk is actually from *not* doing this, because then you’d get reuse of cached‐but‐invalidated FODs.

I was wondering if we should print a “hey, if you just upgraded to 25.05, it’s probably because of the migration, just update the hash” message in the FOD hasher temporarily? But it might be confusing when people just actually forget to update the hash for other reasons.

Reply:
AFAIUI eval warnings still get triggered whether or not a FOD hash is invalidated or not.
So every eval of the package would show this warning.

Reply:
> AFAIUI eval warnings still get triggered whether or not a FOD hash is invalidated or not. So every eval of the package would show this warning.

Only if `useFetchCargoVendor` was explicitly set to false. I'm much more worried about the common case of it not being set at all.

Reply:
24.11 will still need the flag, yes.
When backporting we need to make sure that useFetchCargoVendor = true is still present, otherwise we will get a hash mismatch.

But in case someone forgets about it, it will not go undetected because the derivation name is guaranteed to be different with fetchCargoTarball and fetchCargoVendor, thus it will not blindly fetch from the cache.

Reply:
> AFAIUI eval warnings still get triggered whether or not a FOD hash is invalidated or not. So every eval of the package would show this warning.


Btw, here I was talking about the "warn user about expected has mismatch" idea of Emily.

Reply:
> AFAIUI eval warnings still get triggered whether or not a FOD hash is invalidated or not. So every eval of the package would show this warning.

I’m talking about people upgrading from 24.11 to 25.05 with third‐party packages that have `useFetchCargoVendor` unset. Not about Nixpkgs use. After this PR they won’t get an eval warning but they will get a FOD error. I think that’s probably okay, but we could add a temporary message to the end of the FOD fetcher output if we could think of useful advice to give that won’t be confusing in other situations.

Reply:
Oh, you meant in the builder and not at eval time.

You were thinking about putting the warning into fetchCargoVendor's output logs so when the default changes, and people start unexpectedly using fetchCargoVendor, they will see that it's a known problem.

Putting a warning into fetchCargoTarball as well could also help somewhat when backporting.

Reply:
I think we can’t backport a warning into 24.11 as it isn’t really a backwards‐compatible change. The branch will only be live for three or so more months so hopefully it shouldn’t be a big deal.